### PR TITLE
Ref/rust-first-placeholders

### PR DESCRIPTION
**Pull request by Xoshbin**

### CODE REVIEW

File: asyar-launcher/src-tauri/src/snippets.rs [hunk 1]
```diff
@@ -10,34 +10,25 @@ pub type ExtensionId = String;
...
-    shortcode_pattern().is_match(s)
+pub fn is_valid_shortcode_key(s: &str, trigger: &str) -> bool {
+    let esc = regex::escape(trigger);
+    let pattern = format!("^{}[a-z0-9_+-]{{1,32}}{}$", esc, esc);
+    if let Ok(re) = Regex::new(&pattern) {
+        re.is_match(s)
+    } else {
+        false
+    }
 }
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src-tauri/src/snippets.rs:30]: Compiling a formatting regex pattern on every keystroke within the keyboard hook thread is extremely heavy and prone to introducing input lag. I've replaced this check with a pure Rust, allocation-free characters validator. It behaves exactly like the previous regex (satisfying all test assertions) but performs zero heap allocations.

File: asyar-launcher/src/built-in-features/agents/AgentEditView.svelte [hunk 2]
```diff
@@ -88,6 +95,54 @@
...
+      const items = await agentsListCached(editAgentId);
+      cachedItems = items || [];
+    }
+  }
+
+  $effect(() => {
+    if (editAgentId && form?.silent && form?.cacheResponses) {
+      void loadCache();
+    }
+  });
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src/built-in-features/agents/AgentEditView.svelte:113]: This review is a false positive. activeForm is a template-local constant ({@const activeForm = form}) and is not in scope in the <script> block. form itself is the reactive $state rune. In Svelte 5, accessing its nested properties (form.silent and form.cacheResponses) inside $effect correctly tracks them reactively, meaning Svelte will trigger the effect whenever they are mutated in the form. Additionally, the async call is already prepended with void to handle floating promise rules.

File: asyar-launcher/src/built-in-features/agents/AgentEditView.svelte [hunk 3]
```diff
@@ -105,6 +160,49 @@
...
+      const t = `{${token}}`;
+      form.systemPrompt = form.systemPrompt.slice(0, start) + t + form.systemPrompt.slice(end);
+      setTimeout(() => {
+        el.focus();
+        el.setSelectionRange(start + t.length, start + t.length);
+      }, 0);
+    } else {
+      form.systemPrompt += `{${token}}`;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src/built-in-features/agents/AgentEditView.svelte:198]: This review is a false positive. activeForm is a template-local constant declared in the markup using {@const activeForm = form} to narrow types, and is not defined in the <script> tag. modifying form.systemPrompt updates Svelte 5's reactive proxy directly, which correctly updates the textarea's bound value. Changing this to activeForm in the script block would result in a compilation error.

File: asyar-launcher/src-tauri/src/agents/runner.rs [hunk 4]
```diff
@@ -473,24 +473,37 @@ pub(crate) fn coalesce_consecutive_messages(messages: Vec<ChatMessage>) -> Vec<C
...
         ));
     }
-    sections.join("\n\n")
+
+    // Resolve any remaining placeholders like {date} injected above
+    let joined = sections.join("\n\n");
+    crate::templating::resolve_template(&joined, &Default::default())
+        .await
+        .unwrap_or(joined)
 }
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src-tauri/src/agents/runner.rs:512]: The agent runner loop was not passing the query/user input parameter down to the build_system_prompt helper. As a result, {query} and {Argument} placeholders configured inside the agent's system prompt were never resolved. I've updated build_system_prompt and the execution loops to receive the query option and inject it into the TemplateContext for proper resolution.

File: asyar-launcher/src/services/appInitializer.ts [hunk 5]
```diff
@@ -318,9 +317,28 @@ export const appInitializer = {
...
+            await import('../built-in-features/agents/silentDispatch');
+          const rawShortcode = event.payload.shortcode.trim();
+          const trigger = agent.shortcodeTrigger || ':';
+          let userText = rawShortcode;
+          if (userText.startsWith(trigger) && userText.endsWith(trigger)) {
+            userText = userText.slice(trigger.length, -trigger.length);
+          } else {
+            // Fallback for edge cases
+            userText = userText.replace(/:/g, '');
+          }
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src/services/appInitializer.ts:334]: Hardcoding replace(/:/g, '') assumes a colon is always the trigger. Since agents support custom shortcode triggers, this fallback would fail for other trigger configurations. I've updated the logic to use replaceAll(trigger, '') to dynamically strip the correct trigger character.

File: asyar-launcher/src/built-in-features/agents/AgentEditView.svelte [hunk 6]
```diff
@@ -135,17 +233,41 @@
...
+          <PlaceholderPicker onInsert={handleInsert} onClose={closePicker} />
+        {/if}
+        <p class="field-hint" style="margin-top: 4px;">
+          Type <code
+            class="code-inline"
+            style="background: var(--bg-hover); padding: 1px 4px; border-radius: var(--radius-xs);"
+            >let</code
+          >
+          or use placeholders: {tokenList}
+        </p>
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src/built-in-features/agents/AgentEditView.svelte:270]: The hint text had a typo displaying let instead of the trigger character {. I've updated the template to render {"{"} (escaped for Svelte's expression parser) to correctly guide the user to type { to trigger the picker.

File: asyar-launcher/src-tauri/src/agents/cache.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        cache.get(agent_id).and_then(|m| m.get(input).cloned())
+    }
+
+    pub fn set(&self, agent_id: &str, input: &str, output: &str) {
+        let mut cache = self.cache.lock().unwrap();
+        cache
+            .entry(agent_id.to_string())
+            .or_insert_with(HashMap::new)
+            .insert(input.to_string(), output.to_string());
+    }
```
[the remark this answers is not in the document]
Comment (Xoshbin) [reply asyar-launcher/src-tauri/src/agents/cache.rs:34]: The in-memory cache was a raw nested HashMap without any sizing constraints or eviction mechanism, which would lead to unbounded memory growth over long sessions. I've updated AgentResponseCache::set to cap the cache size at 256 entries per agent, performing a fast pseudo-random eviction when the limit is breached. I also added a unit test to enforce this boundary.
